[TRTLLM-7831][feat] Support block wise FP8 in wide ep

xxi-nv opened the pull request:
<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- New Features
  - Introduced a pluggable MoE backend with automatic runtime selection based on hardware and quantization, improving performance and compatibility.
  - Added a public API to run MoE via the selected backend, with support for fused finalize and tuner controls (token count, top-k).
  - Enhanced FP8 block-wise quantization path with version-aware optimizations.

- Tests
  - Added comprehensive multi-GPU tests validating FP8 block-wise MoE correctness across varying sequence lengths and all-to-all configurations, including cross-checks against a reference implementation.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

<!--
Please write the PR title by following this template:

**[TRTLLM-7831][feat] Support block wise FP8 in wide ep**

Valid ticket formats:
  - JIRA ticket: [TRTLLM-1234] or [FOOBAR-123] for other FOOBAR project
  - NVBugs ID: [https://nvbugs/1234567]
  - GitHub issue: [#1234]
  - No ticket: [None]

Valid types (lowercase): [fix], [feat], [doc], [infra], [chore], etc.

Examples:
  - [TRTLLM-1234][feat] Add new feature
  - [https://nvbugs/1234567][fix] Fix some bugs
  - [#1234][doc] Update documentation
  - [None][chore] Minor clean-up

Alternative (faster) way using CodeRabbit AI:

**[JIRA ticket/NVBugs ID/GitHub issue/None] @coderabbitai title**

NOTE: "@coderabbitai title" will be replaced by the title generated by CodeRabbit AI, that includes the "[type]" and title.
For more info, see /.coderabbit.yaml.

-->

## Description

<!--
New Features
The current wideEP doesn't support FP8 block-wise on BLACKWELL.
This PR refactors the code by introducing MoEDeepGemmBackend, MoECutlassBackend and MoEBackendSelection.
If the wideEP runs on Blackwell with block-wise FP8, then it will choose MoEDeepGemmBackend as the implementation.
-->

## Test Coverage

<!--
test_fused_moe_fp8_blockwise_wide_ep
-->

## PR Checklist

Please review the following before submitting your PR:
- PR description clearly explains what and why. If using CodeRabbit's summary, please make sure it makes sense.
- PR Follows [TRT-LLM CODING GUIDELINES](https://github.com/NVIDIA/TensorRT-LLM/blob/main/CODING_GUIDELINES.md) to the best of your knowledge.
- Test cases are provided for new code paths (see [test instructions](https://github.com/NVIDIA/TensorRT-LLM/tree/main/tests#1-how-does-the-ci-work))
- Any new dependencies have been scanned for license and vulnerabilities
- [CODEOWNERS](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/CODEOWNERS) updated if ownership changes
- Documentation updated as needed
- The reviewers assigned automatically/manually are appropriate for the PR.


- [x] Please check this after reviewing the above items as appropriate for this PR.

## GitHub Bot Help

`/bot [-h] ['run', 'kill', 'skip', 'reuse-pipeline'] ...`

Provide a user friendly way for developers to interact with a Jenkins server.

Run `/bot [-h|--help]` to print this help message.

See details below for each supported subcommand.

<details>

`run  [--reuse-test (optional)pipeline-id --disable-fail-fast --skip-test --stage-list "A10-PyTorch-1, xxx" --gpu-type "A30, H100_PCIe" --test-backend "pytorch, cpp" --add-multi-gpu-test --only-multi-gpu-test --disable-multi-gpu-test --post-merge --extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx" --detailed-log --debug(experimental)]`

Launch build/test pipelines. All previously running jobs will be killed.

`--reuse-test (optional)pipeline-id ` *(OPTIONAL)* : Allow the new pipeline to reuse build artifacts and skip successful test stages from a specified pipeline or the last pipeline if no pipeline-id is indicated. If the Git commit ID has changed, this option will be always ignored. The DEFAULT behavior of the bot is to reuse build artifacts and successful test results from the last pipeline.

`--disable-reuse-test ` *(OPTIONAL)* : Explicitly prevent the pipeline from reusing build artifacts and skipping successful test stages from a previous pipeline. Ensure that all builds and tests are run regardless of previous successes.

`--disable-fail-fast ` *(OPTIONAL)* : Disable fail fast on build/tests/infra failures.

`--skip-test ` *(OPTIONAL)* : Skip all test stages, but still run build stages, package stages and sanity check stages. Note: Does **NOT** update GitHub check status.

`--stage-list "A10-PyTorch-1, xxx"` *(OPTIONAL)* : Only run the specified test stages. Examples: "A10-PyTorch-1, xxx". Note: Does **NOT** update GitHub check status.

`--gpu-type "A30, H100_PCIe"` *(OPTIONAL)* : Only run the test stages on the specified GPU types. Examples: "A30, H100_PCIe". Note: Does **NOT** update GitHub check status.

`--test-backend "pytorch, cpp"` *(OPTIONAL)* : Skip test stages which don't match the specified backends. Only support [pytorch, cpp, tensorrt, triton]. Examples: "pytorch, cpp" (does not run test stages with tensorrt or triton backend). Note: Does **NOT** update GitHub pipeline status.

`--only-multi-gpu-test ` *(OPTIONAL)* : Only run the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--disable-multi-gpu-test ` *(OPTIONAL)* : Disable the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--add-multi-gpu-test ` *(OPTIONAL)* : Force run the multi-GPU tests in addition to running L0 pre-merge pipeline.

`--post-merge ` *(OPTIONAL)* : Run the L0 post-merge pipeline instead of the ordinary L0 pre-merge pipeline.

`--extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx"` *(OPTIONAL)* : Run the ordinary L0 pre-merge pipeline and specified test stages. Examples: --extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx".

`--detailed-log ` *(OPTIONAL)* : Enable flushing out all logs to the Jenkins console. This will significantly increase the log volume and may slow down the job.

`--debug ` *(OPTIONAL)* : **Experimental feature**. Enable access to the CI container for debugging purpose. Note: Specify exactly one stage in the `stage-list` parameter to access the appropriate container environment. Note: Does **NOT** update GitHub check status.

For guidance on mapping tests to stage names, see `docs/source/reference/ci-overview.md`
and the `scripts/test_to_stage_mapping.py` helper.

### kill

`kill  `

Kill all running builds associated with pull request.

### skip

`skip --comment COMMENT `

Skip testing for latest commit on pull request. `--comment "Reason for skipping build/test"` is required. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

### reuse-pipeline

`reuse-pipeline `

Reuse a previous pipeline to validate current commit. This action will also kill all currently running builds associated with the pull request. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

</details>

xxi-nv committed:
feat: wide_ep support block-wise FP8 on blackwell

Signed-off-by: xxi <xxi@nvidia.com>

	modified:   tensorrt_llm/_torch/modules/fused_moe/fused_moe_wide_ep.py
	new file:   tensorrt_llm/_torch/modules/fused_moe/moe_backend.py
	modified:   tests/unittest/_torch/modules/test_fused_moe.py

	modified:   tensorrt_llm/_torch/modules/fused_moe/fused_moe_wide_ep.py
	new file:   tensorrt_llm/_torch/modules/fused_moe/moe_backend.py
	modified:   tests/unittest/_torch/modules/test_fused_moe.py

xxi-nv commented:
This PR depends on https://github.com/NVIDIA/TensorRT-LLM/pull/7422

xxi-nv commented on tensorrt_llm/_torch/modules/fused_moe/moe_backend.py:
```diff
@@ -0,0 +1,791 @@
+        import tensorrt_llm.quantization.utils.fp8_utils as fp8_utils
+        self.fp8_utils = fp8_utils
+
+        from .fused_moe_deepgemm import deepgemm_fp8_group_blockwise_gemm
```
Yeah, that's the idea of this design.
From my perspective, in the future, we could actually select different modules(the MoEBackendImpl) automatically through heuristic selection or autotuning among different tactic providers (MoEBackends).

_Could we merge the DeepGemmFusedMoE and CutlassFusedMoE modules? For example, we could have a single fused module that selects its computational backend at initialization._

QiJune replied:
Thanks for the clarification.

To make this design even clearer and more aligned with standard conventions (`Op` for stateless, `Module` for stateful), what do you think about renaming `MoEBackend` to `MoEOp`?

For example:
- `MoECutlassBackend` -> `CutlassMoEOp`
- `MoEDeepGemmBackend` -> `DeepGemmMoEOp`

This would also lead to a cleaner directory structure:

```python
_torch/modules/fused_moe/
├── ops/
│   ├── moe_op.py
│   ├── moe_op_deepgemm.py
│   └── moe_op_cutlass.py
├── create_moe.py
└── fused_moe_wide_ep.py
```

> I don't want to involve too many refactoring work in this PR, in case it brings potential risks.

Excellent points. Here are a few polished versions of your reply that are more collaborative and persuasive.

Option 1: Collaborative and Persuasive (Recommended)
Thanks for the clarification, I agree with your vision for auto-tuning and heuristic selection of backends.

To make this design even clearer and more aligned with standard conventions (Op for stateless, Module for stateful), what do you think about renaming MoEBackend to MoEOp?

For example:

MoECutlassBackend -> CutlassMoEOp

MoEDeepGemmBackend -> DeepGemmMoEOp

This would also lead to a cleaner directory structure:

Plaintext

_torch/modules/fused_moe/
├── ops/
│   ├── moe_op.py
│   ├── moe_op_deepgemm.py
│   └── moe_op_cutlass.py
├── create_moe.py
└── fused_moe_wide_ep.py
I don't want to involve too many refactoring work in this PR, in case it brings potential risks.

I understand the concern about risk. However, moving `deepgemm_fp8_group_blockwise_gemm` into the `DeepGemmMoEOp` seems like a safe, low-risk change. It would immediately improve code organization by co-locating the function with its backend and removing the local import, which is a bit unconventional.

xxi-nv replied:
1 - If the Op is more consistent with the rule in TRTLLM, I will follow that. 
2 - The refactoring work includes not only moving the `deepgemm_fp8_group_blockwise_gemm `into the `DeepGemmMoEOp`, but also some other utility functions and using the `DeepGemmMoEOp `in `DeepGemmFusedMoE`. 
However, if you think I need to do part of the refactoring, such as moving `deepgemm_fp8_group_blockwise_gemm `into the `DeepGemmMoEOp`, then sure, I can do that.

QiJune replied:
1. Yes, please help rename to `Op`, which is clearer.
2. Please add a TODO in the local import line.

xxi-nv replied:
Sure, thanks for the suggestion.

yuxianq replied:
After reviewed moe_backend.py, I think it is still half-formed now. We still need to:

1. Move some fields/methods to `MoE` to make them public APIs instead of relying on `getattr`
2. Merge `DeepGemmFusedMoE` into `CutlassFusedMoE` and use these MoEOp instead to validate that this design is general enough

I appreciate this modular design, it should be a self-contained refactor PR instead of mixed in this functional PR. Can we simply copy-paste some code from `DeepGemmFusedMoE` to `WideEPMoE` instead in this PR, and make all refactor in a follow-up PR instead?

yuxianq replied:
@litaotju FYI: With this refactor, `DeepGemmFusedMoE` can merge into `CutlassFusedMoE`, so the user don't need to specify `moe_backend: DEEPGEMM` on Blackwell now.

kaiyux commented:
/bot run

tensorrt-cicd commented:
[PR_Github #17258](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/17258/) [ run ] completed with state `SUCCESS`
[/LLM/release-1.1.0rc2/L0_MergeRequest_PR pipeline #7](https://nv/trt-llm-cicd/job/release-1.1.0rc2/job/L0_MergeRequest_PR/7/) completed with status: 'FAILURE'

yuxianq commented on tensorrt_llm/_torch/modules/fused_moe/moe_backend.py:
```diff
@@ -0,0 +1,791 @@
+        getattr(module, 'tune_max_num_tokens', 8192)
+        module.swiglu_alpha
+        module.swiglu_beta
+        module.swiglu_limit
```
All `swiglu_xxx` are unused here.

yuxianq commented on tensorrt_llm/_torch/modules/fused_moe/moe_backend.py:
```diff
@@ -0,0 +1,791 @@
+        cluster_size = module.cluster_size
+        cluster_rank = module.cluster_rank
+        enable_alltoall = module.enable_alltoall
+        getattr(module, 'tune_max_num_tokens', 8192)
```
Don't use unsafe `getattr`, move `tune_max_num_tokens` to MoE instead.

yuxianq commented on tensorrt_llm/_torch/modules/fused_moe/moe_backend.py:
```diff
@@ -0,0 +1,791 @@
+            self,
+            module: 'MoE',  # Now required as first parameter
+            # Input tensors
+        x: torch.Tensor,
```
Indent error here.

yuxianq commented on tensorrt_llm/_torch/modules/fused_moe/moe_backend.py:
```diff
@@ -0,0 +1,791 @@
+
+        Args:
+            module: MoE module containing configuration and parameters.
+                   The following will be extracted:
```
Don't need to record expected fields here, all of them should be a part of `MoE`'s public APIs.

xxi-nv committed:
rename by comment

Signed-off-by: xxi <xxi@nvidia.com>

	modified:   tensorrt_llm/_torch/modules/fused_moe/fused_moe_wide_ep.py
	deleted:    tensorrt_llm/_torch/modules/fused_moe/moe_backend.py
	new file:   tensorrt_llm/_torch/modules/fused_moe/ops/__init__.py
	new file:   tensorrt_llm/_torch/modules/fused_moe/ops/moe_op.py
	new file:   tensorrt_llm/_torch/modules/fused_moe/ops/moe_op_cutlass.py
	new file:   tensorrt_llm/_torch/modules/fused_moe/ops/moe_op_deepgemm.py

xxi-nv commented on tensorrt_llm/_torch/modules/fused_moe/moe_backend.py:
```diff
@@ -0,0 +1,791 @@
+        # Check if we should use DeepGemm backend
+        # Blackwell has SM version 100
+        is_blackwell = get_sm_version() == 100
+        has_block_fp8 = (hasattr(module, 'has_deepseek_fp8_block_scales')
```
Fixed

xxi-nv commented on tensorrt_llm/_torch/modules/fused_moe/moe_backend.py:
```diff
@@ -0,0 +1,791 @@
+        Args:
+            module: The MoE module containing configuration information
+                   Expected attributes:
+                   - has_deepseek_fp8_block_scales: Whether block FP8 is enabled
```
Removed

xxi-nv commented on tensorrt_llm/_torch/modules/fused_moe/moe_backend.py:
```diff
@@ -0,0 +1,791 @@
+        cluster_size = module.cluster_size
+        cluster_rank = module.cluster_rank
+        enable_alltoall = module.enable_alltoall
+        getattr(module, 'tune_max_num_tokens', 8192)
```
Thanks for the suggestion.
The `tune_max_num_tokens ` seems is not used widely by the MoE variants.
Before move it to the MoE, let me do more investigation and leave this refactor work to [TRTLLM-7848](https://jirasw.nvidia.com/browse/TRTLLM-7848)

ziyixiong-nv commented:
/bot run

tensorrt-cicd commented:
[PR_Github #17475](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/17475/) [ run ] triggered by Bot

ziyixiong-nv commented:
/bot run

tensorrt-cicd commented:
[PR_Github #17513](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/17513/) [ run ] triggered by Bot

QiJune reviewed (approved):
LGTM

xxi-nv commented:
/bot run

kaiyux commented:
/bot run --disable-fail-fast

kaiyux commented:
/bot run --disable-fail-fast --add-multi-gpu-test

litaotju commented:
/bot run --disable-fail-fast --add-multi-gpu-test

tensorrt-cicd commented:
[PR_Github #17729](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/17729/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #17698](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/17698/) [ run ] completed with state `ABORTED`

tensorrt-cicd commented:
[PR_Github #17751](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/17751/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #17729](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/17729/) [ run ] completed with state `ABORTED`

tensorrt-cicd commented:
[PR_Github #17751](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/17751/) [ run ] completed with state `SUCCESS`
[/LLM/release-1.1.0rc2/L0_MergeRequest_PR pipeline #76](https://nv/trt-llm-cicd/job/release-1.1.0rc2/job/L0_MergeRequest_PR/76/) completed with status: 'FAILURE'

xxi-nv commented:
/bot run --disable-fail-fast --add-multi-gpu-test

kaiyux commented:
bot run --disable-fail-fast --add-multi-gpu-test

kaiyux commented:
/bot run --disable-fail-fast --add-multi-gpu-test
